Connected Data Platform - Phase 1

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR is the first phase of the CDP. It covers the following tasks:

1. Allows a user to remove trip history.
2. Anonimizes trip data.
3. Uploads zipped trip data to S3.
4. Manages which trip history days should be uploaded to S3.

### New Configuration Parameters

CONNECTED_DATA_PLATFORM_S3_BUCKET_NAME: otp-middleware-dev
CONNECTED_DATA_PLATFORM_S3_FOLDER_NAME: trip-data
CONNECTED_DATA_PLATFORM_TRIP_HISTORY_UPLOAD_JOB_FREQUENCY_IN_MINUTES: 5

### Notes

This update is dependant on the UI providing a batch id. See https://github.com/opentripplanner/otp-react-redux/pull/478

Commit:
refactor(Initial CDP wrk): Initial CDP wrk

Commit:
refactor(Bug fixes): Bug fixes

Commit:
Merge branch 'dev' into conn-data-platform

Commit:
refactor(Addressed bug and created unit tests): Addressed bugs and created unit tests

br648 commented:
@landonreed This is now ready for first review. The tests are failing because GitHub will need the new S3 config params and I'm not sure how to update them. The required trip summary params will need updated. For now, I have focused on getting the whole process in place first.

Commit:
refactor(TripRequest.java): Removed redundant JSON annotation

Commit:
refactor(ConnectedDataPlatformTest.java): Added additional test and renamed class

miles-grant-ibigroup commented on src/main/java/org/opentripplanner/middleware/cdp/AnonymizedTripRequest.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * User Id. {@link OtpUser#id} of user making trip request.
+     */
+    public String userId;
```
Is this removed before being written to a zip file? I'm guessing it's here to allow users to be forgotten, but it does feel strange to have a `userId` field on a class titled `AnonymizedTripRequest`

br648 replied:
This was included in the design doc by Landon. I think the plan was to hold the anonymized data separately and the user id was the key. I've implemented slightly differently (didn't see the need for a separate table) so have now removed from this class.

miles-grant-ibigroup commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,219 @@
+        } finally {
+            // Delete the temporary zip file if it was created. This is done here in case the S3 upload fails.
+            try {
+                if (tempZipFile != null && !IS_TEST) {
```
It might be a good idea to leave a warning in the console that the file isn't actually being deleted when test mode is active?

miles-grant-ibigroup commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Get all dates that have not been uploaded and place in a Set class.
+     */
+    private static Set<Date> getIncompleteUploadsAsSet() {
```
Since this is only used once, would this not be cleaner as an override of `getIncompleteUploads()`?

br648 replied:
Since it is only used once I have removed.

miles-grant-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/TripHistoryUpload.java:
```diff
@@ -0,0 +1,66 @@
+public class TripHistoryUpload extends Model {
+
+    public Date uploadDate;
+    public String status = TripHistoryUploadStatus.PENDING.getValue();;
```
Double semi-colon!

miles-grant-ibigroup commented on src/main/java/org/opentripplanner/middleware/cdp/TripHistoryUploadStatus.java:
```diff
@@ -0,0 +1,26 @@
+        this.value = value;
+    }
+
+    public String getValue() {
```
Is this needed?

br648 replied:
yeah, MongoDB can't save enums. This allows me to save the String value of it.

miles-grant-ibigroup commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,239 @@
+            HttpMethod.PUT
+        );
+        assertEquals(
+            1,
```
We expect this because we've only created one TripHistory, right?

br648 replied:
yeah, I'll add a comment.

miles-grant-ibigroup commented on configurations/default/env.yml.tmp:
```diff
@@ -9,6 +9,10 @@ BUGSNAG_PROJECT_NOTIFIER_API_KEY: project-notifier-api-key
 
+CONNECTED_DATA_PLATFORM_S3_BUCKET_NAME: bucket-name
+CONNECTED_DATA_PLATFORM_S3_FOLDER_NAME: folder-name
+CONNECTED_DATA_PLATFORM_TRIP_HISTORY_UPLOAD_JOB_DELAY_IN_MINUTES: 5
```
Is "delay" the best term here? Isn't it more of a frequency?

br648 replied:
I've used delay in the past, but agree frequency is better, have updated!

miles-grant-ibigroup reviewed (commented):
Looks really good! Just a few small questions here and there about details

Commit:
refactor(Addressed PR feedback): Addressed PR feedback

Commit:
refactor(maven-ci.yml): Added CDP required params

Added CDP required params

Commit:
refactor(maven-ci.yml): Updatd config params

Included CDP params under addition build step

Commit:
refactor(env.schema.json): Added CDP env params

Added CDP specific env params

miles-grant-ibigroup reviewed (approved):
Thanks for those changes! Looks great! Will pass this PR to Evan now.

Commit:
refactor(README.md): Updated to include new CDP params

Updated to include new CDP params

Commit:
refactor(Updated env.schema.json and readme.md): Update includes merged bugsnag and cdp params

Updated to include merged bugsnag and CDP params

Commit:
refactor(Bug fixes): Fixed bugs related to failed unit tests

Fixed bugs related to fialed unit tests

Commit:
refactor(IAM policy details): Update to README and env.yml.tmp to include/reference IAM policy confi

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/AbstractUserController.java:
```diff
@@ -118,6 +119,15 @@ U preCreateHook(U user, Request req) {
+                // files previously uploaded to S3.
+                ConnectedDataManager.removeUsersTripHistory(otpUser.id);
+            }
+        }
```
I was wondering if a user's data should also be removed if they delete their account, but I suppose this approach is best so that users would specifically note that they want to remove their trip history and then they can delete their account. I'm not even sure if it's possible to delete an account tho, lol.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/AnonymizedTripRequest.java:
```diff
@@ -0,0 +1,33 @@
+package org.opentripplanner.middleware.cdp;
```
Please spell out this package name completely.

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/TripRequest.java:
```diff
@@ -82,4 +92,33 @@ public boolean delete() {
+    /**
+     * Get all trip requests between two dates.
+     */
+    private static FindIterable<TripRequest> getTripRequests(Date start, Date end) {
```
It may just be a personal opinion, but since some of these static methods are only ever called once and from the ConnectedDataManager class that perhaps these methods belong in that class instead.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,216 @@
+        for (TripRequest request : TripRequest.requestsForUser(userId)) {
+            userTripDates.add(getStartOfDay(request.dateCreated));
+            // This will delete the trip summaries as well.
+            request.delete();
```
Is there a way to do a bulk-delete instead of a bunch of individual deletes?

br648 replied:
I could go with `removeFiltered(eq("userId", userId))`, but I would still need to loop through each in order to get the trip request id so I could delete the related trip summaries.

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/TripHistoryUpload.java:
```diff
@@ -0,0 +1,66 @@
+
+import java.util.Date;
+
+public class TripHistoryUpload extends Model {
```
Can you add some JavaDoc about what this model is used for?

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/TripHistoryUpload.java:
```diff
@@ -0,0 +1,66 @@
+     * Get all incomplete uploads.
+     */
+    @BsonIgnore
+    public static FindIterable<TripHistoryUpload> getIncompleteUploads() {
```
It'd also be my personal preference to move these static methods into the classes where they're called from.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,216 @@
+     * Process incomplete upload dates. This will be uploads which are flagged as 'pending'. If the upload date is
+     * compiled and uploaded successfully, it is flagged as 'complete'.
+     */
+    public static void processTripHistory() {
+        List<TripHistoryUpload> incompleteUploads = getIncompleteUploads();
+        incompleteUploads.forEach(tripHistoryUpload -> {
+            if (compileAndUploadTripHistory(tripHistoryUpload.uploadDate)) {
+                // Update the status to 'completed' if successfully compiled and uploaded.
+                tripHistoryUpload.status = TripHistoryUploadStatus.COMPLETED.getValue();
+                Persistence.tripHistoryUploads.replace(tripHistoryUpload.id, tripHistoryUpload);
+            }
+        });
+    }
```
It seems like it'd make sense to put this in the `TripHistoryUploadJob`.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,216 @@
+            }
+        });
+        // Kick-off trip data processing to recompile and upload trip data to S3 minus the user's trip history.
+        processTripHistory();
```
This could be a long-running process, so I think it'd be best to start a `TripHistoryUploadJob`.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/AbstractUserController.java:
```diff
@@ -118,6 +119,15 @@ U preCreateHook(U user, Request req) {
+            if(!otpUser.storeTripHistory && existingOtpUser.storeTripHistory) {
+                // If an OTP user no longer wants their trip history stored, remove all history from MongoDB and all
+                // files previously uploaded to S3.
+                ConnectedDataManager.removeUsersTripHistory(otpUser.id);
```
See other related comments about how this could involve starting a long-running process and therefore should trigger the creation of a job.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,216 @@
+        );
+    }
+
+    /**
+     * Add to the upload list any dates between now and the last upload date. This will cover a new day once
+     * passed midnight and any days missed due to downtime.
+     */
+    public static void stageUploadDays() {
+        Date now = new Date();
+        TripHistoryUpload latest = TripHistoryUpload.getLatest();
+        if (latest == null) {
+            // No data held, add the previous day as the first day to be uploaded.
+            Persistence.tripHistoryUploads.create(
+                new TripHistoryUpload(
+                    getStartOfDay(
+                        getDateMinusNumberOfDays(now, 1)
+                    )
+                ));
+        } else {
+            Set<LocalDate> betweenDays = DateTimeUtils.getDatesBetween(
+                getDatePlusNumberOfDays(latest.uploadDate,1),
+                now
+            );
+            betweenDays.forEach(day -> {
+                Persistence.tripHistoryUploads.create(new TripHistoryUpload(getStartOfDay(day)));
+            });
+        }
+    }
```
This would probably also make sense to include in the `TripHistoryUploadJob`.

evansiroky commented on src/main/java/org/opentripplanner/middleware/cdp/ConnectedDataManager.java:
```diff
@@ -0,0 +1,216 @@
+            FileUtils.writeFileToZip(
+                tempFile,
+                getFileName(startOfDay, DATA_FILE_NAME_SUFFIX),
+                getAnonymizedTripHistory(startOfDay, getEndOfDay(date))
```
I'm a little worried that this might create a massive JSON file that would result in an out-of-memory error. Is there a way you can write this data in a streaming fashion?

br648 replied:
Done.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/FileUtils.java:
```diff
@@ -31,4 +42,64 @@ public static String getFileContents(String pathAndFileName) throws IOException
+    /**
+     * Extracts the contents of a file contented within a zip file.
+     */
+    public static String getContentsOfFileInZip(String zipFileNameAndPath, String fileName) throws IOException {
```
It appears this is only used during testing which is good since it attempts to read the entire file into memory which could result in an out-of-memory error if used on a massive zip file. I'd add a warning to the JavaDoc of the class noting this just in case.

evansiroky commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,241 @@
+            tempFile,
+            getFileName(startOfYesterday, ConnectedDataManager.DATA_FILE_NAME_SUFFIX)
+        );
+        String contents = ConnectedDataManager.getAnonymizedTripHistory(startOfYesterday, getEndOfDay(startOfYesterday));
+        assertEquals(fileContents, contents);
```
This doesn't seem to be a great assertion here since a subcomponent of the `processTripHistory` method is being called to generate the expected data. So long as the implementation remains the same, it's always going to match and thus what you're really testing is anything up until the calling of `getAnonymizedTripHistory`. In this case, I'd recommend creating a snapshot of the data to assert against. See example of snapshotting in gtfs-lib: https://github.com/conveyal/gtfs-lib/blob/94ff5fda5785020e3a689e8f5daf33291b4b7625/src/test/java/com/conveyal/gtfs/graphql/GTFSGraphQLTest.java#L99

evansiroky commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,241 @@
+            getFileName(startOfYesterday, ConnectedDataManager.DATA_FILE_NAME_SUFFIX)
+        );
+        TripHistory tripHistory = JsonUtils.getPOJOFromJSON(fileContents, TripHistory.class);
+        assertTrue(tripHistory.tripRequests.stream().anyMatch(tripRequest -> tripRequest.batchId.equals(batchId)));
```
Please add a comment about what this assertion checks.

evansiroky commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,241 @@
+            getFileName(startOfYesterday, ConnectedDataManager.DATA_FILE_NAME_SUFFIX)
+        );
+        tripHistory = JsonUtils.getPOJOFromJSON(fileContents, TripHistory.class);
+        assertTrue(tripHistory.tripRequests.stream().noneMatch(tripRequest -> tripRequest.batchId.equals(batchId)));
```
Please add a comment about what this assertion checks.

evansiroky commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,241 @@
+        TripHistory tripHistory = JsonUtils.getPOJOFromJSON(fileContents, TripHistory.class);
+        assertTrue(tripHistory.tripRequests.stream().anyMatch(tripRequest -> tripRequest.batchId.equals(batchId)));
+        ConnectedDataManager.removeUsersTripHistory(userId);
+        // Trip request and summary are removed as part of the 'removeUsersTripHistory' method. Setting these to null
+        // prevents a delete error under the tidy-up process.
+        tripRequest = null;
+        tripSummary = null;
```
Test cases shouldn't have side effects on other test cases if at all possible. Is there a way these variables can be made to have a scope active only within this method?

evansiroky commented on src/test/java/org/opentripplanner/middleware/cdp/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,241 @@
+            Persistence.tripHistoryUploads.getCountFiltered(Filters.eq("uploadDate", tenDaysInThePast))
+        );
+
+        // Trip request and summary are removed as part of the 'removeUsersTripHistory' method. Setting these to null
+        // prevents a delete error under the tidy-up process.
+        tripRequest = null;
+        tripSummary = null;
+
+        // Set the zip file name and temp file name so they are both removed as part of the tidy-up process.
+        zipFileName = getFileName(tenDaysInThePast, ConnectedDataManager.ZIP_FILE_NAME_SUFFIX);
+        tempFile = String.join(
+            "/",
+            FileUtils.getTempDirectory().getAbsolutePath(),
+            zipFileName
+        );
```
Test cases shouldn't have side effects on other test cases if at all possible. Is there a way these variables can be made to have a scope active only within this method?

evansiroky reviewed (dismissed):
Looks pretty good. Most of my comments are about perhaps personal refactoring preferences. However, the two items that I believe are blocking are:

- try to make writing to the zip file a fully-streamed process
- make it so an API request to remove a user's trip history starts a job that runs independently of the API request.

Commit:
refactor(Address PR feedback): Addressed PR feedback

Commit:
refactor(Removed redundant methods): Removed redundant methods causing test errors

Commit:
refactor(ConnectedDataManager.java): Tweaked object extraction

Tweaked object extraction

Commit:
refactor(Update to include anon transit leg parameters): Includes anon transit leg data and random l

Commit:
refactor(Updates to include a full set of anonymized data): Added anonymized trip request and respon

Commit:
refactor(Bug fixes and addtional testing.): Bug fixes and additional testing.

Commit:
refactor(Updated swagger output): Updated swagger output

Commit:
refactor(Missed completed comment.): Completed missed class comment.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Duration of the trip on this itinerary, in seconds.
+     */
+    public Long duration = 0L;
```
I see that there are both member variables of type `Long` and `long`. Is this one intentionally nullable?

br648 replied:
This was a copy from the original Itinerary class which also has Long and long. I have updated all to Long to be consistent.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * The number of transfers this trip has.
+     */
+    public Integer transfers = 0;
```
Same here. This is nullable. Should it be this way?

br648 replied:
This was a copy from the original Itinerary class which also has Integer. I have left this as is.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Time that the trip departs.
+     */
+    public Date startTime = null;
```
You're introducing `java.util.Date` which is problematic as it has so many deprecated methods, check: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Date.html

It also has a terrible API and John Skeet says ["You should avoid it if you can."](https://codeblog.jonskeet.uk/2017/04/23/all-about-java-util-date/. Luckily it has been replaced with a much nicer one since Java 8: `java.time`

I seriously suggest you should switch to it.

I have quite a bit experience with `java.time` and would love to help if you're unfamiliar with it.

Edit: I see that you're using `java.time` elsewhere in this PR. Why not use `OffsetDateTime` or `Instant` instead of Date?

br648 replied:
I have migrated to java.time where possible.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * From place Lat. Trip starting point.
+     */
+    public Double fromPlaceLat;
```
What's the reason for the lat/long part to be put into separate variables? Would it not make sense to use some type of `Coordinates` object?

br648 replied:
It would yes. I have updated.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:
```diff
@@ -0,0 +1,84 @@
+            this.time = tripRequest.requestParameters.get("time");
+            this.tripType = tripRequest.requestParameters.get("arriveBy");
+            if (tripType != null && tripType.equalsIgnoreCase("true")) {
+                tripType = "Arrive By";
```
Isn't this pretty "stringly typed"? Would you not want to use an enum?

br648 replied:
yep! Have converted to an enum.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,324 @@
+
+    // If set to true, no files are upload to S3 or deleted from the local disk. This is expected to be carried out by
+    // the unit tests instead.
+    public static boolean IS_TEST;
```
I'm not sure that this isn't a code smell. Could it not be replace with a parameter on the method to stop the deletion from happening?

br648 replied:
Yes, have updated.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:
```diff
@@ -0,0 +1,85 @@
+            // No data held, add the previous day as the first day to be uploaded.
+            Persistence.tripHistoryUploads.create(
+                new TripHistoryUpload(
+                    getStartOfDay(
```
This is a good example where `java.time`'s API will make things a lot easier.

br648 replied:
Migrated.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/utils/DateTimeUtils.java:
```diff
@@ -201,4 +207,72 @@ public static ZonedDateTime makeOtpZonedDateTime(String dateString, String timeS
+    /**
+     * Converts a {@link LocalDateTime} object into a {@link Date} object using the Otp zone id.
+     */
+    public static Date convertToDate(LocalDateTime dateToConvert) {
```
I'm repeating myself but I would use a `OffsetDateTime` or `ZonedDateTime`.

leonardehrenfried replied:
This will make lots of these utiity methods redundant as they are so much easier to use.

br648 replied:
Migrated.

leonardehrenfried commented on src/main/java/org/opentripplanner/middleware/utils/LatLongUtils.java:
```diff
@@ -0,0 +1,105 @@
+ * be found and tested here: http://jsfiddle.net/hoolymama/56wzdtax/.
+ */
+public class LatLongUtils {
+    public static boolean IS_TEST = false;
```
Again, I'm not too hot about this. I would create some form of randomiser interface or method that has two implementations: one for production and one for testing.

br648 replied:
Fair point. I have overloaded the method.

leonardehrenfried reviewed (changes requested):
Looks good, but I have a few questions.

The only thing that I really care about though is not using `java.util.Date`.

Commit:
refactor(Addressd PR feedback): Addressed Pr feedback

br648 commented:
@leonardehrenfried I have addressed your PR feedback minus anything related to Date/Time. This will be covered in tonights (05/10) tech call. I will address according to feedback from that.

Commit:
refactor(BugsnagTest): Updated assertion so test will not be effected by what time of day it is run

Commit:
refactor(Migrated away from java.util.Date): Replaced java.util.Date where possible with LocalDate

Commit:
refactor(PersistenceTestUtils.java): Bug fix preventing unit test from succeeding

Commit:
refactor(Additional test to cover trip summary with error): New test as well as date/time bug fix

Commit:
refactor(ConnectedDataPlatformTest): Added back in end to end test

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * How much time is spent walking, in seconds.
+     */
+    public long walkTime = 0;
```
Make this one nullable too?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java:
```diff
@@ -0,0 +1,57 @@
+
+/**
+ * Anonymous version of {@link org.opentripplanner.middleware.otp.response.Itinerary} containing only parameters
+ * flagged as anonymous.
```
Could you clarify "flagged"? Actually, could you clarify "anonymized"? Maybe a better wording could be "parameters that don't contain precise user or location data".  (Also update JavaDoc description of other anonymized classes that have a similar description.)

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedPlace.java:
```diff
@@ -0,0 +1,22 @@
+
+    public Date arrival;
+    public Date departure;
+    public Coordinates coordinates;
```
Add a comment saying the coordinates will be scrambled upon writing to a file.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:
```diff
@@ -0,0 +1,73 @@
+        Coordinates toCoordinates
+    ) {
+        this.batchId = tripRequest.batchId;
+        this.fromPlace = fromCoordinates;
```
Remove duplicate statement.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    public static void stageUploadDays() {
+        LocalDate now = LocalDate.now();
+        TripHistoryUpload latest = TripHistoryUpload.getLatest();
```
This should probably be `getLatestCompleted` (in JavaDoc too). The current check does not look for complete status of upload.

br648 replied:
I'm looking for the last trip history upload regardless of status as a starting point for staging upload days see `TripHistoryUploadJob.stageUploadDays()`. I've updated the method to `getLastCreated` and updated the Javadoc accordingly.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * Add to the upload list any dates between now and the last upload date. This will cover a new day once
```
Description is incorrect. Update description to reflect that the code triggers upload of trip requests from the last upload date up to the day before now.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+     * 2) For each batch id get matching trip requests.
+     * 3) Workout which trip request uses the most modes.
+     * 4) Define lat/lon for 'from' and 'to' places.
+     * 5) Anonymize trip request with the most modes.
```
Add to this line: ", scrambling location coordinates for non-public locations"

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+        boolean isTest
+    ) throws IOException {
+        // Get distinct batchId values between two dates.
+        DistinctIterable<String> uniqueBatchIds = Persistence.tripRequests.getDistinctFieldValues(
```
The UI currently sends out all requests with batch IDs set to -1, so another string needs to be used for uniqueness.

br648 replied:
This was discussed at the initial design meeting with Landon. The UI would be updated to provide a unique id and not -1. I don't think there is another param that can be used?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+
+            // Get all trip summaries matching the batch id.
+            FindIterable<TripSummary> tripSummaries = Persistence.tripSummaries.getFiltered(
+                eq("batchId", uniqueBatchId),
```
Trip summaries should be matched to one of the fetched trip requests above by `tripRequestId`.

br648 replied:
See previous comment. I think this needs to remain batch id.

binh-dam-ibigroup replied:
You are right on this one. A change in https://github.com/opentripplanner/otp-react-redux/pull/329 will cause only one trip request to be recorded, but trip summaries for trip requests under the same batch will still be recorded and have no parent `tripRequestId`... In that case could we remove the `tripRequestId` field from the `TripSummary` class?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    /**
+     * Obtain anonymize trip data for the given date, write to zip file, upload the zip file to S3 and finally delete
```
Typo "anonymized"

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/AbstractUserController.java:
```diff
@@ -118,6 +120,18 @@ U preCreateHook(U user, Request req) {
+            OtpUser otpUser = (OtpUser) user;
+            OtpUser existingOtpUser = (OtpUser) preExistingUser;
+            if(!otpUser.storeTripHistory && existingOtpUser.storeTripHistory) {
+                // If an OTP user no longer wants their trip history stored, remove all history from MongoDB.
```
Do we need to show a confirmation message about that in the UI?

br648 replied:
Yes, probably prior to the user hitting send/update. This can be done with the batch id update.

br648 replied:
Addressed here: https://github.com/opentripplanner/otp-react-redux/pull/478

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+            }
+        }
+
+        String coords = place.split("::")[1].trim();
```
Careful, the UI might send just the coordinates (if the geocoder does not return anything, which is unlikely).
If that happens, the format will just be `lat,lon` and `::` will not be present.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    /**
+     * Extract the trip request that uses the most modes.
```
Expand on the reasoning behind this.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/TripRequest.java:
```diff
@@ -36,24 +38,27 @@
-    //TODO: This could be the request parameters returned as part of the plan response. Would be POJO based instead of just text.
-    public String queryParams;
+    /** A dictionary of the parameters provided in the request that triggered this response. */
+    public HashMap<String, String> requestParameters;
```
Note for self: Ooooh this change needs to be reflected in https://github.com/opentripplanner/otp-react-redux/pull/329.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/FileUtils.java:
```diff
@@ -31,4 +41,82 @@ public static String getFileContents(String pathAndFileName) throws IOException
+    }
+
+    /**
+     * Extracts the contents of a file contented within a zip file. This method will read the entire contents of a file
```
Replace "contented" with "contained".

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataPlatformTest.java:
```diff
@@ -0,0 +1,426 @@
+        List<AnonymizedTrip> anonymizedTrips = JsonUtils.getPOJOFromJSONAsList(fileContents, AnonymizedTrip.class);
+        assertNotNull(anonymizedTrips);
+        assertEquals(LatLongUtils.TEST_LAT, anonymizedTrips.get(0).tripRequest.fromPlace.lat);
+        assertEquals(LatLongUtils.TEST_LON, anonymizedTrips.get(0).tripRequest.fromPlace.lon);
```
You assert using `Coordinates.equals`.

binh-dam-ibigroup reviewed (changes requested):
It looks like some changes in the logic for compiling trips and for determining the last complete S3 update are necessary. See also other changes.

Commit:
refactor(Addressed PR feedback): Addressed PR feedback

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -232,21 +233,22 @@ private static Coordinates getPlaceCoordinates(
     /**
-     * Extract the trip request that uses the most modes.
+     * A single trip query results in many calls from the UI to OTP covering different combinations of modes. The trip
+     * request to be included in the anonymous trip data must be the one which uses the most modes.
```
This method could probably deleted in favor of retrieving the first request found for a batch. We don't use the number of modes involved for anything, do we?

br648 replied:
@binh-dam-ibigroup this method only exists so that the trip request with the most modes is serialized and uploaded to S3. This way the client can see the maximum number of modes that were used to produce the trip summaries. It is possible that I haven't understood this correctly, if so, let me know.

br648 replied:
Addressed after tech call and code update.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:
```diff
@@ -22,20 +22,21 @@ public void run() {
-        if (latest == null) {
+        TripHistoryUpload lastCreated = TripHistoryUpload.getLastCreated();
+        if (lastCreated == null) {
             // No data held, add the previous day as the first day to be uploaded.
```
Should the same logic from the else block be applied on first run? (upload the last say 20 days?)

br648 replied:
No, this is to cover the case where the job has not run before e.g. initial deployment.

binh-dam-ibigroup reviewed (changes requested):
Looks mostly good, but remove method `getTripRequestWithMostModes` because we don't consider the number of modes anywhere else for this PR.

Commit:
refactor(ConnectedDataManager.java): Added Bugsnag exception reporting

Added Bugsnag exdeption reporting to catch trip data processing failures

Commit:
refactor(OtpRequestProcessor.java): Fixed merge conflict

Fixed merge conflict

Commit:
refactor(Update to capture all modes used in a batch): The anonymized trip request now contains all

Commit:
refactor(ConnectedDataPlatformTest.java): Corrected encoded comma typo

Commit:
refactor(Corrected snapshots): Updated snapshots to include comma separated modes

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -234,21 +239,27 @@ private static Coordinates getPlaceCoordinates(
             }
+            Set<String> uniqueModesForRequest = new HashSet<>(
+                // Expecting the comma separating modes to be encoded as '%2C'.
+                Arrays.asList(tripRequest.requestParameters.get("mode").split("%2C"))
```
You could just add this to `allUniqueModes`.

br648 replied:
Good spot! Updated.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:
```diff
@@ -0,0 +1,350 @@
+            }
+            Set<String> uniqueModesForRequest = new HashSet<>(
+                // Expecting the comma separating modes to be encoded as '%2C'.
+                Arrays.asList(tripRequest.requestParameters.get("mode").split("%2C"))
```
This assumption is incorrect. It looks like `TripRequest` objects are created in `OtpRequestProcessor` using the OTP response (and not the URL query params), where the mode value is in the form below, not encoded: 
```js
requestParameters: {
  ...
  mode: "CAR_PARK,WALK,TRANSIT",
  ...
}
```

binh-dam-ibigroup replied:
Also, the values appear to be stored that way in Mongo too, so there is no `%2C` in the modes extracted from Mongo.

br648 replied:
Thanks. Updated.

binh-dam-ibigroup reviewed (changes requested):
Please address https://github.com/ibi-group/otp-middleware/pull/152/files#r738677278.

Commit:
refactor(Corrected encoded comma references): Replace encoded comma use with normal commas

binh-dam-ibigroup reviewed (approved):
Looks good and seems to work now! Thanks for the changes.

Commit:
refactor(Fixed merge conflicts): Fixed merge conflicts with dev branch
